# Large-Neighbourhood-Search für das Traveling-Salesman-Problem

bearbeitet von Markus Lang, Seminar Analytics & Optimization: Methods & Software, Tag der Abgabe: 10.07.2020

## 1. Einleitung
Das Traveling Salesman Problem ist ein mittlerweile breit erforschtes kombinatorisches Optimierungsproblem. Es ist dafür bekannt, trotz seiner Einfachheit mit wachsender Modellinstanz sehr schnell sehr komplex zu werden. Als kleines Rechenbeispiel ist hier ein Postbote zu nennen, welcher Briefe an insgesamt 15 verschiedene Haushalte austrägt. Der Postbote muss an jedem Briefkasten vorbeikommen, um seine Briefe auszuliefern. Für die Wahl der Reihenfolge stehen nun 15! verschiedene Möglichkeiten, also insgesamt über 1.3 Billionen verschiedene Möglichkeiten, zur Auswahl. Das Phänomen, welches dahinter steckt, wird auch als kombinatorische Explosion bezeichnet. Mit jeder zusätzlichen Erweiterung des Modells, in unserem Fall jeder zusätzliche Briefkasten, wird das Modell noch viel komplexer. Ganz zu schweigen von der Betrachtung eines Tourenplanungsproblems, bei welchem noch weitere Restriktionen wie die Fahrzeugkapazität, festgelegte Auslieferzeiten und Dienstpausen sowie komplexe Lohnzahlungsverfahren hinzukommen. Da so ein Modell mit optimierenden Verfahren nur bis zu einer gewissen Modellgröße sinnvoll lösbar ist, und die Probleme in der realen Welt deutlich größer und komplexer sind, bieten sich hierbei heuristische Verfahren zur Anwendung auf Tourenplanungsprobleme als Lösung an (vgl. Shaw (1998, S. 417f)).
Ein heuristisches Verfahren, welches im Rahmen dieser Arbeit betrachtet werden soll, ist die Large Neighbourhood Search. Dieses eignet sich besonders für die Anwendung auf große und stark eingeschränkte Tourenplanungsprobleme. Eine Heuristik erlangt zwar nicht zwangsläufig die beste, also die optimale Lösung, besticht allerdings durch eine vergleichsweise einfache Implementierung, eine schnelle Berechnung und eine gute Lösungsqualität. 
Zunächst sollen im Folgenden die theoretischen Hintergründe einer Large Neighbourhood Search erklärt werden, wobei auf den generellen Aufbau und Ablauf des Verfahrens eingegangen wird. Anschließend folgt eine Implementierung der LNS anhand von zwei eigens erstellten Beispielen, wobei auf den Code allgemein und die Besonderheiten der Umsetzung eingegangen wird. Ein kurzer Schlussteil rundet schließlich die Arbeit ab.

## 2. Theoretische Grundlagen
### 2.1 Beschreibung des Traveling-Salesman-Problems
Das Traveling Salesman Problem, kurz TSP, ist ein kombinatorisches Optimierungsproblem eines Handlungsreisenden. Der Handlungsreisende muss dabei alle Städte in seiner Umgebung bedienen. Innerhalb einer Tour, welche in einem bestimmten Knoten startet, werden also alle Knoten des Modells vom Handlungsreisenden einmal besucht, bevor die Tour wieder im Startknoten des Modells endet. Der direkte Weg zwischen zwei Knoten wird durch eine Kante gekennzeichnet, wobei für jede Kante die entsprechenden Kosten $c_{ij}$ ermittelt werden, welche anfallen, um von Knoten $i$ zu Knoten $j$ zu gelangen. Die Kosten zwischen den jeweiligen Knoten sind dabei in beide Richtungen gleich, es liegt folglich ein symmetrisches TSP vor. Das Ziel dieses Problems ist nun, eine Rundreise für den Handlungsreisenden so zu konstruieren, dass alle Knoten des Modells genau einmal besucht werden und die Gesamtkosten der Tour dabei so klein wie möglich gehalten werden. Für die Zielfunktion des Modells bedeutet das, dass die Summe aller vom Handlungsreisenden genutzten Kanten $c_{ij}$ minimiert werden soll. Die Gesamtkosten der Tour sind dabei zum Einen von der Entscheidung welche Kanten genutzt werden sollen abhängig, zum Anderen von der Wahl der Reihenfolge der Knoten innerhalb der Tour (vgl. Pisinger und Ropke (2010, S. 400)).

### 2.2 Aufbau und Ablauf einer LNS
Das Grundprinzip des Large-Neighbourhood-Search-Verfahrens besteht darin, eine bereits gefundene zulässige Lösung noch weiter zu verbessern. Hierbei wird die Nachbarschaft einer Lösung nach weiteren Verbesserungen abgesucht. Die direkte Nachbarschaft einer untersuchten Lösung besteht hierbei aus grundsätzlich ähnlichen Lösungen, welche durch eine kleine Veränderung der aktuellen Lösung erreicht werden können. Im Rahmen der Large Neighborhood Search, kurz LNS, geschieht dies, indem gewisse Teile einer bereits vorliegenden Lösung erst zerstört und anschließend wieder zu einer neuen zulässigen Lösung zusammengefügt werden. Hierfür werden sowohl Destroy- als auch Repair-Operatoren benötigt, welche durch ihr Zusammenwirken nach jedem Verfahrensablauf eine neue Lösung kreieren (vgl. Pisinger und Ropke (2010, S. 406ff)). Die untersuchten Nachbarschaften werden also durch Veränderungen an der aktuellen Lösung im Rahmen eines Ruin-and-Recreate-Schrittes erreicht.
Bei den Destroy-Operatoren stehen verschiedenste Möglichkeiten zur Auswahl. Es können hierbei beispielsweise eine zufällige Anzahl an Knoten, ein bestimmter geografischer Bereich oder eine bestimmte Abfolge aus Knoten aus dem aktuellen Tourenplan einer zulässigen Lösung entfernt werden. Was all diese Verfahren im Grunde gemeinsam haben, ist dass sie ein gewisses Zufallselement enthalten. Hierbei wird sichergestellt, dass auch alle verschiedensten Bereiche einer vorliegenden Lösung beachtet werden und auf Verbesserungen untersucht werden können. Auch die Entscheidung, wie groß der Anteil der Zerstörung der aktuellen Lösung ist, spielt eine bedeutende Rolle. Generell sollte der Anteil der Veränderung nicht zu klein gewählt werden, da ansonsten zwangsläufig auch die Nachbarschaft, welche von der aktuellen Lösung aus erreicht werden kann, sehr klein gehalten wird. Eine LNS-Heuristik macht sich jedoch genau dies zunutze und liefert durch eine solche Einschränkung möglicherweise nicht die bestmöglichen Ergebnisse. Ein allzu hoher Zerstörungsanteil sollte allerdings auch vermieden werden, da ansonsten der Repair-Prozess wiederum sehr komplex werden kann, wodurch es schwerer wird, eine neue bessere Lösung zu finden (vgl. Schrimpf et al. (2000, S. 147f)).
Auch auf Seiten des Repair-Operators gibt es unterschiedliche Vorgehensweisen. Hier können sowohl heuristische Verfahren, bei denen beispielsweise die entfernten Knoten unter Berücksichtigung der geringsten Zusatzkosten in den Tourenplan wieder eingefügt werden, als auch optimierende Verfahren, welche die entfernten Knoten unter Berücksichtigung einer zu minimierenden Zielfunktion wieder einfügt, verwendet werden. Der Vorteil in der Verwendung heuristischer Verfahren besteht hierbei, dass sie schneller sind als optimierende Verfahren, aber dennoch eine gute Lösungsqualität liefern.
Eine LNS-Metaheuristik navigiert sich durch das abwechselnde Zusammenspiel von Destroy- und Repair-Operatoren durch den Lösungsraum, bis ein Abbruchkriterium erreicht wird. Hierbei können sowohl ein Iterationslimit als auch ein Zeitlimit verwendet werden. Der Umgang mit den gefundenen Lösungen bietet ebenfalls Spielraum für große Variationen im Verfahrensablauf. Es kann zum Beispiel grundsätzlich jede gefundene Lösung akzeptiert werden, wobei hier auch Verschlechterungen des Zielfunktionswertes in Kauf genommen werden. In jedem Iterationsschritt wird dabei einfach die gefundene Lösung der vorherigen Iteration weiterverwendet und mithilfe des Destroy- und Repair-Operators im darauf folgenden Iterationsschritt verändert. Die beste gefundene Lösung wird während des Verfahrensablaufs gespeichert und gegebenenfalls aktualisiert, falls eine bessere Lösung gefunden wurde. Nach Erreichen des Abbruchkriteriums kann so die beste gefundene Lösung einfach abgerufen werden. Diese Vorgehensweise der erlaubten Verschlechterung verhindert ein mögliches Feststecken der LNS-Metaheuristik in einem lokalen Optimum. Des Weiteren können auch gewisse Toleranzkriterien festgelegt werden, welche eine Verschlechterung der Lösung gegenüber der vorherigen Lösung bis zu einem gewissen Grad zulässt.  Verschlechternde Lösungen können auch mit einer vorher festgelegten Wahrscheinlichkeit akzeptiert werden. Im Gegensatz dazu können allerdings auch Restriktionen festgelegt werden, durch welche nur Lösungen berücksichtigt werden, die auch wirklich eine Verbesserung oder zumindest keine Verschlechterung gegenüber der vorherhigen Lösung erwirken (vgl. Logistics Management (2019, S.121)).

### 2.3 Abgrenzung zu weiteren Heuristiken
Eine Adaptive Large Neighborhood Search, kurz ALNS, unterscheidet sich von der LNS-Heuristik hinsichtlich der Wahl der Operatoren. Während bei der LNS der Destroy- und der Repair-Operator bereits im Vorhinein festgelegt werden muss, lässt eine ALNS mehrere Operatoren zu. Es werden lediglich Gewichte bzw. Wahrscheinlichkeiten festgelegt, mit der ein jeweiliger Operator verwendet wird. Das bedeutet, dass sich die Operatoren innerhalb des Verfahrensablaufs von Iteration zu Iteration unterscheiden können. Während bei der LNS sich für jeweils einen Operator entschieden werden muss, welcher allgemein gute Lösungen finden kann und somit generell gut auf viele verschiedene Probleminstanzen angewendet werden kann, ermöglicht eine ALNS-Heuristik auch die Einbeziehung spezieller Operatoren. Diese Operatoren sind allerdings weniger gut zur allgemeinen Verwendung auf unterschiedliche Probleminstanzen geeignet. Durch die Verwendung unterschiedlicher Heuristiken läuft eine ALNS seltener in Gefahr, in einem lokalen Optimum festzustecken, und ermöglicht gleichzeitig eine Berücksichtigung mehrerer unterschiedlicher Nachbarschaften, da diese implizit von der Wahl der Operatoren abhängen (vgl. Pisinger und Ropke (2010, S. 409ff)). 
Die Variable Neighborhood Search, kurz VNS, macht sich wie die ALNS mehrere Nachbarschaften zunutze. Dabei wird eine Nachbarschaft mit variabler Tiefe genutzt. Falls eine VNS sich in einem lokalen Optimum einer Nachbarschaft befindet, wird die variable Struktur der Nachbarschaft genutzt und auf eine größere Nachbarschaft gewechselt. Falls dies gelingt, wechselt die VNS-Heuristik wieder auf eine kleinere Nachbarschaft. Im Gegensatz dazu arbeitet eine LNS-Heuristik nur mit einer Nachbarschaft, und diese ist wie bereits gesagt durch die Wahl der Operatoren im Vorhinein festgelegt (vgl. Pisinger und Ropke (2010, S. 414)).
Bei traditionellen Local-Search-Heuristiken wie beispielsweise eine Tabuliste oder ein Or-Opt-Operator ist eine breite Untersuchung der Nachbarschaften und des Lösungsraumes abhängig von der gewählten Heuristik und somit nur bedingt möglich. LNS-Verfahren bieten im Gegensatz dazu die Möglichkeit, durch die Verwendung von Zufallselementen im Destroy- und im Repair-Operator eine größere Diversifikation in der Erforschung des Lösungsraumes zu erhalten. Stagnierende beziehungsweise sich wiederholende Veränderungen an der Lösungsinstanz können so unterbunden werden und ermöglichen daraus folgend einen besseren Zugang zu mehreren Nachbarschaften und einen größeren Lösungsraum. Ein weiterer Unerschied besteht in der Wahl der Nachbarschaften. Während diese bei einer lokalen Suche expizit ausgewählt werden muss, ist diese Entscheidung bei der LNS implizit gegeben (vgl. Pisinger und Ropke (2010, S.412ff)). Des Weiteren unterbinden lokale Suchverfahren durch das Zulassen von Verschlechterungen ein mögliches Feststecken in lokalen Optima und vergrößern somit die Wahrscheinlichkeit, ein globales Optimum zu erreichen. Generell lässt sich jedoch behaupten, dass ein LNS-Verfahren besser mit vielen Nebenbedingungen umgehen kann als ein Local-Search-Verfahren, wodurch es sich auch besser zur Anwendung auf realitätsnähere Tourenplanungsprobleme eignet. Ein Auftreten mehrerer Nebenbedingungen schränkt nämlich die Anzahl der zulässigen Züge ein, welche im Rahmen eines lokalen Suchverfahrens zur Lösungsverbesserung durchgeführt werden. (vgl. Schrimpf et al. (2000, S.423)).

## 3. Implementierung der LNS-Heurisitik anhand von Beispielinstanzen

Im Folgenden soll nun eine mögliche Umsetzung des LNS-Verfahrens anhand von selbst erstellten Beispielinstanzen gezeigt werden. Hierbei wird zuerst eine LNS anhand einer kleinen TSP-Instanz implementiert, um die Abläufe des LNS-Verfahrens besser schildern zu können. Im Anschluss folgt eine Umsetzung der LNS für ein größeres Capacitated Vehicle Routing Problem, kurz CVRP, welches aufgrund seiner Größe und Komplexität ein deutlich realitätsnäheres Modell darstellt. Die Umsetzungen der Modelle erfolgen hierbei in Form eines Jupyter Notebooks, wobei für den Programmiercode die Programmiersprache Python verwerndet wurde. Im Rahmen der folgenden Unterkapitel wird auch auf mögliche Vorteile und Probleme der verwendeten Programme eingegangen. Als Destroy-Operatoren wurden das Random Ruin- und das Radial Ruin-Verfahren verwendet (vgl. Schrimpf et al. (2000, S. 147f)). Als Repair-Operator und als Eröffnungsverfahren wurde eine Parallel Insertion-Heuristik verwendet (vgl. Logistics Management (2019, S. 113ff)).

### 3.1 Veranschaulichung der LNS anhand eines TSP

Zunächst steht ein einfaches TSP im Vordergrund. Es liegen hierbei 20 Kunden vor, welche vom Handlungsreisenden bedient und somit im Rahmen einer Tour jeweils einmal besucht werden müssen. Der Startpunkt des Handlungsreisenden liegt im Depot bei den Koordinaten (0, 0), während die x- und y-Koordinaten der Kunden zufällig verteilt im Raum (-100, 100) liegen. Das Setzen eines Random-Seeds gewährleistet, dass bei jedem erneuten Durchlauf dieselben Zufallszahlen verwendet werden. Dies stellt sicher, dass sich die Ergebnisse der Umsetzung bei einem erneuten Durchlaufen nicht verändern und somit besser nachvollzogen werden können. Das Ziel des TSP besteht darin, alle 21 Knoten des Modells so in den Tourenplan zu integrieren, dass dieser kostenminimal wird. Die Funktion "distanzberechnung" ermittelt die Länge jeder möglichen Teistrecke zwischen zwei Knoten des Modells, indem die euklidische Distanz zwischen den Knoten $i$ und $j$ ermittelt wird und dieser Wert an den Funktionsaufruf zurückgegeben wird. Der Funktionsaufruf erfolgt, nachdem zur Erstellung der Kostenmatrix $c(i,j)$ ein leeres Array angelegt wurde, und bewirkt, dass dieses Array mit den euklidischen Distanzen befüllt wird. Die Distanz- bwz. Kostenmatrix $c(i,j)$ speichert somit die Entfernungen von einem Knoten $i$ zu einem anderen Knoten $j$. Zu beachten ist, dass die Entfernung zwischen 2 Knoten in beide Richtungen gleich groß ist und folglich ein symmetrisches TSP vorliegt. Außerdem wird durch die exakte Berechnung der Entfernungen aus den Koordinaten der einzelnen Knoten sichergestellt, dass die Dreiecksungleichung automatisch erfüllt ist. Die Dreiecksungleichung besagt, dass der direkte Weg zwischen zwei Knoten immer der kürzeste ist und folglich auch kürzer als ein Umweg über einen dritten Knoten. Zuletzt wird noch ein Big-M, also ein hinreichend großer Wert, welcher die Implementierung vereinfacht, festgelegt.

In [1]:
#Import-Statements
import random
import numpy as np
import math
import copy

#Erstellung der Knotenlisten des Modells
knoten = list(range(21))
anzKnoten = len(knoten)
kunden = knoten[1:]
anzKunden = len(kunden)

#Generierung zufällig verteilter Knoten
#das Setzen eines seeds gewährleistet gleiche Verwendung der Zufallsvariablen bei einer Wiederholung des Modells 
#und gewährleistet somit Nachvollziehbarkeit
random.seed(42)
#Setzen des Depots in die Mitte des Koordinatensystems
koordinatenDepot = [(0, 0)]
#erstelle 20 zufallsbasierte Knoten
koordinatenKunden = [(random.randint(-100, 100), random.randint(-100, 100)) for i in range(1, 21)]
#Koordinatenliste aller Knoten des Modells
koordinatenKnoten = koordinatenDepot + koordinatenKunden

#Funktion zur Distanzberechnung zwischen den einzelnen Knoten
def distanzberechnung(a,b):
    dx = a[0] - b[0]
    dy = a[1] - b[1]
    return (math.sqrt(dx*dx + dy*dy))

#Erstelung und anschließende Berechnung der Distanz- bzw. Kostenmatrix c
c = np.zeros((anzKnoten, anzKnoten), dtype=float)
for i in range(anzKnoten):
    for j in range(anzKnoten):
        c[i,j] = distanzberechnung(koordinatenKnoten[i], koordinatenKnoten[j])

#Big-M
M = 1000000

#wird für spätere Iterationen gebraucht
kundenOffen = copy.deepcopy(kunden)

Nachdem die ersten Rahmenbedingungen der Implementierung festgelegt wurden, kann nun mit der Schilderung der Parallel Insertion-Heuristik begonnen werden. Bei der Verwendung dieser Insertion-Heuristik zur Erstellung einer zulässigen Startlösung muss zunächst eine Initialisierung stattfinden, bevor mit den ersten Iterationsschritten begonnen werden kann. Nachdem ein neuer leerer Tourenplan mit dem Depot als Start- und Endpunkt der Tour angelegt wurde, wird eine Funktion zur Ermittlung des Seed-Customers erstellt. Der Seed-Customer ist hierbei der Kunde, welcher noch nicht in einen Tourenplan eingefügt wurde und aktuell am weitesten vom Depot entfernt ist. Der Start der Insertion-Heuristik mit einem Seed-Customer ist notwendig, da ansonsten immer jeweils der nächste Nachbar des Depots in den Tourenplan eingefügt wird, da dieser die geringsten Insertion Costs verursacht. Eine Berücksichtigung von Seed-Customern liefert folglich meist bessere Ergebnisse. Die Funtion "seedcustomer" kopiert die Depotzeile $c(0,j)$ der Kostenmatrix $c(i,j)$ und speichert diese in einer neuen Liste. Die Liste "depotZeile" gibt nun die Entfernungen aller Kunden $j$ vom Depot an und im Anschluss kann der am Weitesten entfernte Kunde per argmax-Funktion gefunden und als Seed-Customer gekennzeichnet werden. Anschließend wird der Seed-Customer in den Tourenplan an mittlerer Stelle eingefügt und aus der Liste der offenen Kunden, also der Kunden, die noch nicht in den Tourenplan eingefügt worden sind, entfernt.

In [2]:
#erstelle leeren Tourenplan
tourenplan = [0,0]

#Funktion zur Ermittlung des Seed-Customers
def seedcustomer():
    depotZeile = copy.deepcopy(c[0,:])
    sc = np.argmax(depotZeile)
    return sc

#Aufruf der seedcustomer-Funktion
u = seedcustomer()
#Entfernen des Seed-Customers aus der Liste der offenen Kunden
kundenOffen.remove(u)
#Einfügen des Seed-Customers in den Tourenplan
tourenplan.insert(1,u)

#wird später benötigt
tourenplanSeed = copy.deepcopy(tourenplan)
kundenOffenSeed = copy.deepcopy(kundenOffen)

Nun kann eine zulässige Startlösung durch Verwenden der Parallel Insertion-Heuristik konstruiert werden. Hierbei müssen zunächst weitere Import-Statements getätigt werden, welche für die grafische Darstellung der Ergebnisse und eine Interaktion mit diesen notwendig sind. Im Anschluss wird eine Funktion "insertion" für die Parallel Insertion-Heuristik erstellt. Nachdem der Funktion die Ergebnisse des Initialisierungsschrittes übergeben werden, wird in jedem Iterationsschritt der for-Schleife ein neuer Kunde in den Tourenplan eingefügt. Pro Iterationsschritt werden dabei zuerst die Insertion Costs auf eine hinreichend große Zahl festgesetzt. Im Anschluss wird ein jeder offener Kunde $j$ betrachtet und dessen Insertion Costs für jede mögliche Einfügestelle in der Tour ermittelt. Als Einfügestellen sind hierbei alle Stellen der aktuellen Tour außer die erste und letzte Stelle zulässig, da das Depot der Start- und Endpunkt des Tourenplans sein muss. Falls die aktuell gefundenen Insertion Costs kleiner als die bisher besten gefundenen Insertion Costs sind, werden die aktuellen Insertion Costs als neue beste Lösung gespeichert und sowohl die entsprechende Einfügestelle als auch der jeweilige Kunde $j$ festgehalten. Nachdem die beiden inneren for-Schleifen durchgelaufen sind, wurde der Kunde $j$ mit den geringsten Insertion Costs gefunden und dieser kann nun in den Tourenplan an der vorgesehenen besten Einfügestelle $m$ in den Tourenplan aufgenommen werden. Im Anschluss wird der Kunde aus der Liste der offenen Kunden entfernt und ein neuer Iterationsablauf bestimmt den nächsten einzufügenden Kunden. Nachdem alle Kunden in den Tourenplan aufgenommen wurden, kann mit der Visualisierung der Ergebnisse begonnen werden. Hierfür werden zwei Listen erstellt, welche mit den x- und y-Koordinaten des Tourenplans befüllt werden. Falls Kunden noch nicht eingefügt worden sind, werden diese in anderen Listen gespeichert und zur besseren Nachvollziehbarkeit auch mit ausgegeben.
Das Jupyter Notebook bietet durch das Hnzufügen interaktiver Widgets eine tolle Möglichkeit, die Ergebnisse beziehungsweise die Vorgehensweise der Parallel Insertion-Heuristik besser nachvollziehen zu können. Durch das Hinzufügen eines Sliders kann die Iteration, welche man betrachten möchte, festgelegt und somit der gesamte Verfahrens- bzw. Iterationsablauf Schritt für Schritt ausgegeben werden. Dies geschieht durch eine interaktive Verknüpfung des Sliders mit der "insertion"-Funktion. Im ausgegebenen Plot können nun die einzelnen Ergebnisse der jeweiligen Iterationsschritte betrachtet werden. Die Kunden werden dabei mit roten Punkten und das Depot mit einem blauen Kreuz gekennzeichnet. Ein Setzen des Iterationssliders auf den Wert 0 stellt die Initialisierung mit dem eingefügten Seed-Customer dar, während Iteration 19 die Startlösung abbildet.

In [3]:
#Import-Statements:
%matplotlib inline
import matplotlib.pyplot as plt
from ipywidgets import interact, interactive, fixed, interact_manual
import ipywidgets as widgets

#Insertion-Heuristik für das TSP:
def insertion(it):
    tourenplan = copy.deepcopy(tourenplanSeed)
    kundenOffen = copy.deepcopy(kundenOffenSeed)
    #iterationsweiser Ablauf der Parallel Insertion-Heuristik
    for i in range(it):
        pOpt = M
        for j in kundenOffen:
            for m in range(1, len(tourenplan)):
                if c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1],tourenplan[m]] < pOpt:
                    mOpt = m
                    jOpt = j
                    pOpt = c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1], tourenplan[m]]
        tourenplan.insert(mOpt,jOpt)
        kundenOffen.remove(jOpt)
    #grafische Darstellung der Ergebnisse
    x = []
    y = []
    for kunde in tourenplan:
        x.append(koordinatenKnoten[kunde][0])
        y.append(koordinatenKnoten[kunde][1])
    xRest = []
    yRest = []
    for i in kundenOffen:
        xRest.append(koordinatenKnoten[i][0])
        yRest.append(koordinatenKnoten[i][1])
    plt.plot(x, y, 'r.-', 0, 0, 'bx', xRest, yRest, 'r.')
    #aktueller Zielfunktionswert
    zf = 0
    for j in tourenplan[:-1]:
        nachfolgerJ = tourenplan[tourenplan.index(j)+1]
        zf += c[j,nachfolgerJ]
    print("Zielfunktionswert nach aktueller Iteration:",zf)
    
#Hinzufügen interaktiver Widgets
iterationSlider = widgets.IntSlider(value=0, min=0, max=19, step=1, description="Iteration")
widgets.interact(insertion, it=iterationSlider);

interactive(children=(IntSlider(value=0, description='Iteration', max=19), Output()), _dom_classes=('widget-in…

Nachdem die Parallel Insertion-Heuristik als Eröffnungsverfahren anschaulich erläutert wurde, kann nun die gefundene Startlösung mithilfe eines Detroy-Operators teilweise zerstört und anschließend mittels eines Repair-Operators wieder zusammengefügt werden. Hierzu wird zuerst eine neue Funktion definiert. Um den Destroy-Schritt vollziehen zu können, muss zuerst eine Anzahl an Kunden ermittelt werden, welche aus dem aktuellen Tourenplan, beziehungsweise in diesem Fall aus der Startlösung,  entfernt werden. Hierfür kann der Anteil $f$ der aus der aktuellen Tour zu entfernenden Kunden per Schieberegler festgelegt werden. Die Anzahl $a$ aller zu entfernenden Kunden berechnet sich dann durch Multiplikation des Anteils $f$ mit der Gesamtzahl aller Kunden aus dem Modell. Im Anschluss wird eine leere Liste $B$ erstellt, welche später mit den konkreten zu entfernenden Knoten befüllt wird. Da im Rahmen dieser Implementierung zwei Destroy-Operatoren gegenübergestellt werden sollen, dient eine if-Bedingung zur Fallunterscheidung, welcher Operator verwendet werden soll. Dieser kann unten im Output per Widget ausgewählt werden. Beim Random Ruin-Operator wird zunächst eine neue Liste aller Kunden des Modells angelegt, welche die Anzahl der Kunden, die sich in der aktuellen Tour befinden, abbilden soll. Aus dieser Liste wird ein zufälliger Kunde ausgewählt und im Anschluss aus dem Tourenplan entfernt und an die Liste $B$ angehängt. Dies wird im Rahmen einer for-Schleife so lange wiederholt, bis die Anzahl aller entfernten Kunden der vorher festgelegten Anzahl aller zu entfernenden Kunden $a$ entspricht. Beim Radial Ruin wird nur ein Kunde aus der Tour zufällig ausgewählt. Zur Betimmung der weiteren zu entfernenden Kunden wird eine Kopie der Kostenmatrixzeile $c$ des entsprechenden Kunden in einer neuen Liste gespeichert und dessen nächste Nachbarn durch eine argmin-Funktion bestimmt. Ein Gleichsetzen des ersten Eintrags der Liste mit einem Big-M verhindert, dass fälschlicherweise das Depot als zu entfernter Knoten ermittelt wird, da ja nur Kunden aus der aktuellen Tour entfernt werden sollen. Sobald der nächste Nachbar bestimmt wurde, wird dessen entsprechender Eintrag in der Liste mit einem Big-M festgelegt, um zu verhindern, dass derselbe Kunde im nächsten Durchlauf der for-Schleife erneut ausgewählt wird und wirklich $a$ verschiedene Kunden aus dem Tourenplan entfernt werden. Zuletzt werden die entfernten Kunden in die Liste $B$ eingefügt. 
Im Anschluss muss eine zulässige Startlösung erneut ermittelt werden, da die Ergebnisse der vorherigen Code-Zelle leider nicht übernommen werden konnten. Ein weiterer Widget-Slider ermöglicht eine interaktive Ansicht der einzelnen Verfahrensschritte. Falls nur die Startlösung angezeigt werden soll, werden die Ergebnisse der Startlösung ausgegeben und die in $B$ festgelegten Knoten schwarz markiert. Hier kann nun beobachtet werden, wie unterschiedlich bei beiden Destroy-Operatoren die zu entfernenden Knoten ausgewählt werden. Falls der Destroy- oder der Repair-Schritt angezeigt werden soll, geschieht erst die tatsächliche Entfernung der in $B$ festgelegten zu entfernenden Kunden aus den Tourenplan. Hierbei ermöglicht die äußere for-Schleife eine Betrachtung aller in $B$ befindenden Kunden und stellt so sicher, dass auch wirklich alle entsprechenden Kunden entfernt werden. Die innere Schleife vergleicht den kompletten Tourenplan mit dem jeweiligen Kunden aus $B$ und entfernt diesen, sobald er gefunden wurde. Der nun teilweise zerstörte Tourenplan wird im Anschluss ausgegeben, falls der Destroy-Schritt zur Betrachtung ausgewählt wurde. Ein anschließender Repair-Operator tritt erst nach Auswahl des Repair-Schrittes in Kraft. Die Parallel Insertion-Heuristik funktioniert hierbei als Repair-Operator fast identisch wie als Generator einer Startlösung, allerdings wird hier zum Einen kein Seed-Customer benötigt, zum Anderen läuft die äußere for-Schleife so lange durch, bis auch wirklich alle im vorherigen Schritt entfernten Knoten wieder in den Tourenplan eingefügt wurden. Zuletzt werden diese Ergebnisse wieder grafisch ausgegeben.
Mithilfe der interaktiven Widgets kann im Folgenden der genaue Ablauf eines Ruin and Recreate-Schrittes nachvollzogen und die Unterschiede der beiden Destroy-Operatoren herausgestellt werden.

In [4]:
def DestroyAndRepair(f,schritt,auswahl):
    random.seed(42)
    #Anzahl zufällig ausgewählter Knoten, die aus dem Tourenplan entfernt werden:
    a = int(f * anzKunden)
    B = []
    #Entscheidung, welcher Destroy-Operator gewählt werden soll
    if auswahl == "Random Ruin":
        #Random-Ruin:
        kundenInTour = copy.deepcopy(kunden)
        for i in range(a):
            entfernterKunde = random.choice(kundenInTour)
            kundenInTour.remove(entfernterKunde)
            B.append(entfernterKunde)
    else:
        #Radial-Ruin:
        kundenInTour = copy.deepcopy(kunden)
        zufälligerKunde = random.choice(kundenInTour)
        nachbarschaftKunde = copy.deepcopy(c[zufälligerKunde,:])
        nachbarschaftKunde[0] = M
        for i in range(a):
            entfernterKunde = np.argmin(nachbarschaftKunde)
            nachbarschaftKunde[entfernterKunde] = M
            B.append(entfernterKunde)
    #Finden einer zulässigen Startlösung
    tourenplan = copy.deepcopy(tourenplanSeed)
    kundenOffen = copy.deepcopy(kundenOffenSeed)
    while kundenOffen != []:
        pOpt = M
        for j in kundenOffen:
            for m in range(1, len(tourenplan)):
                if c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1],tourenplan[m]] < pOpt:
                    mOpt = m
                    jOpt = j
                    pOpt = c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1], tourenplan[m]]
        tourenplan.insert(mOpt,jOpt)
        kundenOffen.remove(jOpt)
    if schritt == "Startlösung":
        #Plot der Startlösung
        x = []
        y = []
        for kunde in tourenplan:
            x.append(koordinatenKnoten[kunde][0])
            y.append(koordinatenKnoten[kunde][1])
        xB = []
        yB = []
        for i in B:
            xB.append(koordinatenKnoten[i][0])
            yB.append(koordinatenKnoten[i][1])
        plt.plot(x, y, 'r.-', 0, 0, 'bx', xB, yB, 'ko')
    if schritt == "Destroy" or schritt == "Repair":
        #Tourenplan nach Entfernen der Kunden aus B aus dem aktuellen Tourenplan
        for i in B:
            for j in tourenplan:
                if i == j:
                    tourenplan.remove(i)
        if schritt == "Destroy":
            #Plot nach Destroy
            x = []
            y = []
            for kunde in tourenplan:
                x.append(koordinatenKnoten[kunde][0])
                y.append(koordinatenKnoten[kunde][1])
            xB = []
            yB = []
            for i in B:
                xB.append(koordinatenKnoten[i][0])
                yB.append(koordinatenKnoten[i][1])
            plt.plot(x, y, 'r.-', 0, 0, 'bx', xB, yB, 'ko')
    if schritt == "Repair":
        #Repair
        while B != []:
            pOpt = M
            for j in B:
                for m in range(1, len(tourenplan)):
                    if c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1],tourenplan[m]] < pOpt:
                        mOpt = m
                        jOpt = j
                        pOpt = c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1], tourenplan[m]]
            tourenplan.insert(mOpt,jOpt)
            B.remove(jOpt)
        #Plot nach Repair
        x = []   #Tourenplan in x-Koordinaten übersetzen
        y = []   #Tourenplan in y-Koordinaten übersetzen
        for kunde in tourenplan:
            x.append(koordinatenKnoten[kunde][0])
            y.append(koordinatenKnoten[kunde][1])
        plt.plot(x, y, 'r.-', 0, 0, 'bx')

#Auswahl des Anteils der zu entfernenden Kunden
fractionSlider = widgets.FloatSlider(value=0.2, min=0, max=1, step=0.05, description="Anteil der zu entfernenden Knoten:")
#Auswahl des Iterationsschrittes
schrittSlider =  widgets.SelectionSlider(options=["Startlösung", "Destroy", "Repair"], value="Startlösung", 
                                         description="Schritt:")
#Auswahl des Destroy-Operators
auswahlDestroy = widgets.ToggleButtons(options=["Random Ruin", "Radial Ruin"], description="Destroy-Operator:")
widgets.interact(DestroyAndRepair, f=fractionSlider, schritt=schrittSlider, auswahl=auswahlDestroy);

interactive(children=(FloatSlider(value=0.2, description='Anteil der zu entfernenden Knoten:', max=1.0, step=0…

Nachdem die einzelnen Schritte einer LNS anschaulich erläutert wurden, soll nun ein größerer Rundlauf für das vorliegende TSP im Rahmen von 100 Iterationsschritten erfolgen. Dafür wird die Funktion "tsp100" erstellt, welcher zu Beginn die Ergebnisse des Seed-Customer-Iterationsschrittes übergeben werden. Im Anschluss muss dann erneut die zulässige Startlösung ermittelt werden, da diese wie bereits gesagt der Funktion nicht übergeben werden kann. Im Anschluss soll der Zielfunktionswert der Startlösung berechnet werden. Dies geschieht, indem der ZF-Wert zunächst auf 0 festgesetzt wird und für jede im Rahmen der Tour genutzte Teilstrecke mit den entsprechend anfallenden Kosten erhöht wird. Eine for-Schleife stellt dabei sicher, dass jeder Kunde des Tourenplans betrachtet wird. Für jeden Knoten wird dabei sein direkter Nachfolger, also der Knoten, welcher im Tourenplan auf den betrachteten Knoten folgt, ermittelt und der entsprechende Eintrag in der $c$-Matrix herausgesucht. Die for-Schleife endet bereits beim vorletzten Knoten des Tourenplans, da der letzte Knoten, das Depot, keinen Nachfolger besitzt. Danach werden der beste gefundene ZF-Wert und sein entsprechender Tourenplan sowie die entsprechende Iteration, in der die Lösung gefunden wurde, separat gespeichert. Diese Werte gilt es im Rahmen der Iterationsabläufe zu verbessern. Der weitere Iterationsverlauf ähnelt dem bereits genannten aus der Betrachtung des Destroy and Repair-Prozesses weiter oben, jedoch mit dem Unterschied, dass eine neue äußere for-Schleife festgelegt wird, welche das Durchlaufen der Iterationen ermöglicht. Die Anzahl der Iterationen kann mithilfe des entsprechenden Widgets beeinflusst werden. Eine grafische Ausgabe findet nur noch ganz am Ende des Codes statt, da hier nur noch das genaue Endergebnis einer jeweiligen Iteration betrachtet werden soll. Auf die genaueren Vorgänge innerhalb eines Iterationsschrittes wurde bereits weiter oben eingegangen. Am Ende einer jeden Iteration wird der Zielfunktionswert der aktuellen Lösung ermittelt und mit der bisher besten gefundenen Lösung verglichen. Falls ein neuer bester ZF-Wert gefunden wurde, wird der entsprechende Wert mit dem aktuell gefundenen ZF-Wert aktualisiert. Die Iteration des besten gefundenen ZF-Wertes muss dabei um eins erhöht werden, da dies mit der Zählweise des range-Befehles zusammenhängt. Dieser beginnt nämlich bei null, und die nullte Iteration ist im vorliegenden Fall die Startlösung, ab Iteration eins werden erstmals die Destroy- und Repair-Operatoren angewendet. Zuletzt werden die Ergebnisse grafisch ausgegeben, wodurch der Einfluss von Veränderungen der vorliegenden Parameter visuell nachvollzogen werden können.

In [5]:
def tsp100(f, iteration, auswahl):
    #Übergeben der Ergebnisse aus der Seed-Customer-Berechnung
    tourenplan = copy.deepcopy(tourenplanSeed)
    kundenOffen = copy.deepcopy(kundenOffenSeed)
    #Ermittlung der Startlösung durch Insertion-Heuristik
    while kundenOffen != []:
        pOpt = M
        for j in kundenOffen:
            for m in range(1, len(tourenplan)):
                if c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1],tourenplan[m]] < pOpt:
                    mOpt = m
                    jOpt = j
                    pOpt = c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1], tourenplan[m]]
        tourenplan.insert(mOpt,jOpt)
        kundenOffen.remove(jOpt)
    #Ermittlung der Zielfunktion der Startlösung
    zf = 0
    for j in tourenplan[:-1]:
        nachfolgerJ = tourenplan[tourenplan.index(j)+1]
        zf += c[j,nachfolgerJ]
    zfBest = zf
    tourenplanBest = tourenplan
    itBest = 0
    #Anzahl zufällig ausgewählter Knoten, die aus dem Tourenplan entfernt werden
    a = int(f * anzKunden)
    B = []
    #zur Nachvollziehbarkeit der Ergebnisse in anschließender Iterationsschleife
    random.seed(42)
    #Beginn der Iterationsschleife
    for it in range(iteration):
        #Auswahl des Destroy-Operators
        if auswahl == "Random Ruin":
            #Random-Ruin:
            kundenInTour = copy.deepcopy(kunden)
            for i in range(a):
                entfernterKunde = random.choice(kundenInTour)
                kundenInTour.remove(entfernterKunde)
                B.append(entfernterKunde)
        else:
            #Radial-Ruin:
            kundenInTour = copy.deepcopy(kunden)
            zufälligerKunde = random.choice(kundenInTour)
            nachbarschaftKunde = copy.deepcopy(c[zufälligerKunde,:])
            nachbarschaftKunde[0] = M
            for i in range(a):
                entfernterKunde = np.argmin(nachbarschaftKunde)
                nachbarschaftKunde[entfernterKunde] = M
                B.append(entfernterKunde)
        #Destroy
        for i in B:
            for j in tourenplan:
                if i == j:
                    tourenplan.remove(i)
        #Repair
        while B != []:
            pOpt = M
            for j in B:
                for m in range(1, len(tourenplan)):
                    if c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1],tourenplan[m]] < pOpt:
                        mOpt = m
                        jOpt = j
                        pOpt = c[tourenplan[m-1],j] + c[j,tourenplan[m]] - c[tourenplan[m-1], tourenplan[m]]
            tourenplan.insert(mOpt,jOpt)
            B.remove(jOpt)
        #Ermittlung der Zielfunktion der aktuellen Iteration:
        zf = 0
        for j in tourenplan[:-1]:
            nachfolgerJ = tourenplan[tourenplan.index(j)+1]
            zf += c[j,nachfolgerJ]
        #Prüfen, ob neuer bester ZF-Wert gefunden wurde
        if zf < zfBest:
            tourenplanOpt = tourenplan
            zfBest = zf
            itBest = it + 1
    #Plotten der Ergebnisse
    x = []   #Tourenplan in x-Koordinaten übersetzen
    y = []   #Tourenplan in y-Koordinaten übersetzen
    for kunde in tourenplan:
        x.append(koordinatenKnoten[kunde][0])
        y.append(koordinatenKnoten[kunde][1])
    plt.plot(x, y, 'r.-', 0, 0, 'bx')
    print("Lösung der aktuellen Iteration:")
    print("aktueller ZF-Wert:", zf)
    print("aktueller Tourenplan:", tourenplan)
    
    print("optimale gefundene Lösung:")
    print("optimaler ZF-Wert:", zfBest)
    print("bester gefundener Tourenplan:", tourenplanBest)
    print("Iteration, in der die beste Lösung gefunden wurde:", itBest)
    
fractionSlider100 = widgets.FloatSlider(value=0.2, min=0, max=1, step=0.05, description="Anteil der zu entfernenden Knoten:", 
                                        continuous_update=False)
iterationSlider100 = widgets.IntSlider(value=100, min=0, max=100, step=1, description="Iteration", continuous_update=False)
auswahlDestroy100 = widgets.ToggleButtons(options=["Random Ruin", "Radial Ruin"], description="Destroy-Operator:", 
                                          continuous_update=False)

widgets.interact(tsp100, f=fractionSlider100, iteration=iterationSlider100, auswahl=auswahlDestroy100);

interactive(children=(FloatSlider(value=0.2, continuous_update=False, description='Anteil der zu entfernenden …

## 3.2 Umsetzung der LNS für ein CVRP
Als letztes soll nun noch eine Umsetzung der LNS für ein CVRP gezeigt werden, wobei im Unterschied zu vorher jetzt ein Problem mit 100 Kunden und kundenindividuellen Nachfragen vorliegt. Aufgrund seiner Struktur ist dieses Modell deutlich realitätsnäher als das vorherige TSP, weshalb sich auch die Auswirkungen einzelner Parameter des LNS-Verfahrens besser analysieren lassen.
Die Einführung der Modelldaten ist hier grundsätzlich zu der in Kapitel 3.1 ähnlich, jedoch sollen nun 100 geografisch zufällig verteilte Kunden erstellt werden. Auf die Distanzberechung und die Kostenmatrix $c$ wurde bereits eingegangen. Zudem ist eine Einführung der notwendigen Import-Statements hier überflüssig, da diese bereits weiter oben getätigt wurden. Allerdings muss darauf geachtet werden, dass die Codezeilen des Notebooks von oben nach unten ausgeführt werden, damit die Import-Statements ausgeführt werden können. 
Im Unterschied zum TSP liegen im Folgenden den Kunden des Modells nun individuelle Nachfragen zugrunde. Hierfür wird für jeden Kunden eine Zufallszahl aus dem Intervall [1, 10] gezogen und in einer neuen Liste abgespeichert. Die Nachfrage des Depots liegt bei 0. Im Anschluss gibt ein einfacher Print-Befehl die Summe der gesamten Nachfrage aus. Dies wird später für die Einstellung eines interativen Fahrzeugkapazitäts-Parameters benötigt, da man bei Betrachtung der gesamten Nachfrage sehen kann, ab welchem Wert für das vorliegende Modell die Kapazitätsbeschränkung in Kraft tritt. Ein weiterer Unterschied besteht in der Anzahl der Fahrzeuge. Während im vorherigen TSP nur eine gesamte Tour erstellt und im Tourenplan abgespeichert wurde, stehen beim VRP nun mehrere Fahrzeuge, im vorliegenden Fall 3, zur Auswahl. Die Zählweise der Fahrzeugliste beginnt hierbei bei 0 und nicht bei 1, da die Zählweise der Programmiersprache Python auch bei 0 beginnt und es somit vor allem später beim Einfügen der Kunden in den Tourenplan zu Problemen kommen kann.

In [6]:
#Erstellen der Knoten- und Kundenlisten
knoten = list(range(101))
anzKnoten = len(knoten)
kunden = knoten[1:]
anzKunden = len(kunden)

#Setzen des Random-Seeds
random.seed(4)
#Koordinaten des Depots
koordinatenDepot = [(0, 0)]
#erstelle 100 zufallsbasierte Knoten
koordinatenKunden = [(random.randint(-100, 100), random.randint(-100, 100)) for i in range(100)]
#Koordinatenliste aller Knoten des Modells
koordinatenKnoten = koordinatenDepot + koordinatenKunden

#Funktion zur Berechnung der euklidischen Distanz zwischen zwei Knoten
def distanzberechnung(a,b):
    dx = a[0] - b[0]
    dy = a[1] - b[1]
    return (math.sqrt(dx*dx + dy*dy))

#Distanz- bzw. Kostenmatrix c
c = np.zeros((anzKnoten, anzKnoten), dtype=float)
for i in range(anzKnoten):
    for j in range(anzKnoten):
        c[i,j] = distanzberechnung(koordinatenKnoten[i], koordinatenKnoten[j])
        
#Nachfrage des Depots
nachfrageDepot = [0]
#Ermittlung zufälliger Nachfragen aller Kunden
nachfrageKunden = [random.randint(1, 10) for i in range(anzKunden)]
#Liste der Nachfragen aller Knoten
nachfrage = nachfrageDepot + nachfrageKunden

#Ausgabe der gesamten Nachfrage (später für Festlegung der Fahrzeugkapazität relevant)
print("Gesamte Nachfrage aller Kunden:", sum(nachfrage))

#Big-M
M = 1000000

#Anzahl der Fahrzeuge
fahrzeuge = [0, 1, 2]

#wird später im Iterationsablauf benötigt
kundenOffen = copy.deepcopy(kunden)

Gesamte Nachfrage aller Kunden: 532


Zur Ermittlung des Tourenplans aller Fahrzeuge des Modells soll nun ein neuer leerer Tourenplan angelegt werden. Dieser dient im Unterschied zum vorherigen TSP als eine Art Sammlung aller Touren der jeweiligen Fahrzeuge und enthält somit nun mehrere Touren. Die Funktion zur Ermittlung des Seed-Customers ist der aus dem vorherigen Kapitel identisch. Was sich allerdings geändert hat, ist der Funktionsaufruf. Dieser wird nun im Rahmen einer for-Schleife pro Fahrzeug jeweils einmal aufgerufen und bewirkt somit, dass für jede Tour des gesamten Tourenplans ein Seed-Customer ermittelt und in die jeweilige Tour integriert wird. Für jedes Fahrzeug sprechen wir hierbei die jeweilige Tour des Fahrzeugs im Tourenplan durch die Variable $k$ an. Ein großer Nachteil der Parallel Insertion-Heuristik bei der Verwendung als Eröffnungsverfahren ist, dass im Vorhinein durch die Initialisierung aller Touren mit einem Seed-Customer zwingend alle Fahrzeuge des Fuhrparks eingesetzt werden müssen. Zudem muss die Anzahl der Fahrzeuge im Vorhinein festgelegt werden, man muss also schon vorher abschätzen können, wie viele Fahrzeuge zur Bedienung aller Kunden insgesamt benötigt werden.

In [8]:
#Funktion zur Ermittlung des Seed-Customers:
def seedcustomer():
    depotZeile = copy.deepcopy(c[0,:])
    eingefügteKunden = list(set(knoten) - set(kundenOffen))
    for i in eingefügteKunden:
        depotZeile[i] = -M
    sc = np.argmax(depotZeile)
    return sc

#Ermittlung der Startlösung
#erstelle K leere Touren
tourenplan = []
for k in fahrzeuge:
    tourenplan.append([0,0])
#Ermittlung des Seed-Customers für jede Tour k:
for k in fahrzeuge:
    u = seedcustomer()
    kundenOffen.remove(u)
    tourenplan[k].insert(1,u)
tourenplanSeed = copy.deepcopy(tourenplan)
kundenOffenSeed = copy.deepcopy(kundenOffen)

Zu Beginn der Funktion wird die Startlösung ermittelt. Im Unterschied zur vorherigen Implementierung wird in der Insertion-Heuristik eine weitere Restriktion eingeführt. Die Kapazitätsrestriktion ermittelt, ob ein weiterer Kunde in eine Tour eingefügt werden darf. Die Summe aller Nachfragen aller Kunden einer Tour darf dabei nicht die Ladekapazität eines einzelnen Fahrzeugs überschreiten. Der restliche Teil der Insertion-Heuristik bleibt identisch. Nach der Ermittlung des Zielfunktionswertes der Startlösung wird diese als bisher bester gefundener Wert abgespeichert. Eine weitere for-Schleife stellt hier sicher, dass für alle im Tourenplan enthaltenen Fahrzeugtouren auch wirklich alle anfallenden Kosten erfasst werden. Die for-Schleife für den Ablauf der Iterationen beginnt wie bereits gewohnt. Ein erster Unterschied lässt sich jedoch bei der Anwendung des Destroy-Schritts feststellen, denn hier stellt eine weitere for-Schleife sicher, dass über alle Touren des gesamten Tourenplans iteriert wird. Die Repair-Heuristik enthält ebenfalls wieder die Kapazitätsrestriktion für die Fahrzeuge. Im Anschluss wird der Zielfunktionswert der nach einem Destroy & Repair-Schritt gefundenen Lösung ermittelt und mit der bisher besten gefundenen Lösung verglichen. Falls durch den aktuellen ZF-Wert keine Verbesserung erzielt werden kann, wird entschieden, ob der aktuelle Tourenplan trotzdem weiterverwendet werden darf oder verworfen wird. Hierfür wird eine Liste erstellt und mit mehreren "Ja"- und "Nein"-Werten befüllt. Die Anzahl der "Ja"'s kann durch den Slider bestimmt werden. Der Rest der Liste wird mit "Nein"'s befüllt. Falls beispielsweise 0.05 als Auswahlwahrscheinlichkeit gewählt wird, wird die entsprechende Liste mit einem "Ja" und 19 "Nein"-Werten befüllt. Ein anschließend zufällig aus der Liste ausgewählter Wert hat somit mit der Wahrscheinlichkeit 1 zu 20, also 0.05, den Wert "Ja". Falls ein "Ja" aus der Liste gezogen wird, darf die aktuell gefundene Lösung trotz einer Verschlechterung beibehalten werden. Falls ein "Nein" gezogen wird, wird der aktuell gefundene Tourenplan verworfen und in der nächsten Iteration mit der bisher besten gefundenen Lösung weitergerechnet. Eine Modellierung auf diese Weise ermöglicht es, gleichzeitig mehrere Auswahlstrategien in einem abzubilden. Falls nämlich durch das entsprechende Widget die Wahrscheinlichkeit 0 festgelegt wird, wird auch kein "Ja"-Wert aus der Liste gezogen werden, da keiner enthalten ist. Das bedeutet, dass die anschließende if-Bedingung immer erfüllt ist und somit jeder Tourenplan, der zu keiner Verbesserung im ZF-Wert führt, verworfen wird. Dies entspricht der Strategie, dass nur Tourenpläne, welche auch wirkliche ZF-Verbesserungen bringen, weiterverwendet werden. Falls die Wahrscheinlichkeit jedoch auf 1 festgelegt wird, bedeutet das, dass die Liste nur aus "Ja"-Werten besteht und die anschließende if-Bedingung niemals erfüllt sein wird. Es wird folglich kein einziger gefundener Tourenplan mehr abgelehnt. Dies entspricht der Strategie, dass auch Verschlechterungen im ZF-Wert in Kauf genommen und die entsprechenden Tourenpläne trotzdem beibehalten werden. Hierdurch besteht die Möglichkeit, aus einem lokalen Optimum ausbrechen zu können, falls das LNS-Verfahren sich in einem befinden sollte. Eine Wahl der Wahrscheinlichkeit zwischen 0 und 1 bedeutet, dass verschlechternde Ergebnisse mit der festgelegten Wahrscheinlichkeit beibehalten werden. Falls ein neuer bester Zielfunktionswert in der aktuellen Iteration gefunden wurde, wird der entsprechende ZF-Wert, der aktuelle Tourenplan und die Iteration, in der dieser gefunden wurde, abgespeichert. Zudem ist darauf zu achten, dass zum Iterationswert eine 1 zusätzlich hinzuaddiert werden muss, da die Zählweise ansonsten falsch wäre. Python beginnt nämlich immer von 0 zu zählen, jedoch ist im vorliegenden Fall die 0-te Iteration die Startlösung nach Anwendung des Eröffnungsverfahrens. Zuletzt werden die berechneten Ergebnisse ausgegeben.
Mithilfe der Widgets im Output können nun Veränderungen gewisser Parameter vorgenommen und dessen Effekte erforscht werden. Im grafischen Output befindet sich das Ergebnis nach Ende der aktuellen Iteration, welche am entsprechenden Schiebe-Regler eingestellt wurde. Falls der Regler der Auswahlwahrscheinlichkeit kleiner als 1 gewählt wurde, kann es passieren, dass Ergebnisse, welche den ZF-Wert nicht verbessern, verworfen werden. In diesem Fall bildet der grafische Output den bisher besten gefundenen Tourenplan ab, da der Tourenplan der aktuellen Iteration ja verworfen wurde. Deshalb wird mithilfe eines Print-Befehls für jedes Iterationsergebnis vermerkt, ob die Verschlechterung verworfen oder beibehalten wurde, oder ob gar ein neuer bester ZF-Wert in der aktuellen Iteration gefunden wurde. Es kann folglich veranschaulicht werden, welchen Effekt die gewählte Strategie des Umgangs mit Verschlechterungen auf den besten gefundenen ZF-Wert hat. Gleichzeitig kann der Anteil der Zerstörung, also wie viele Knoten aus dem aktuellen Tourenplan entfernt werden sollen, verändert werden. Hierbei ist jedoch zu beachten, dass ein Anteil über 0.5 bereits deutlich längere Berechnungszeiten benötigt als kleinere Zerstörungs-Anteile. Dies beruht darauf, dass ein jeder weiterer Kunde, der im Anschluss des Destroy-Schrittes auch wieder in den Tourenplan integriert werden muss, den Insertion-Prozess deutlich komplexer macht. Ein hoher Destroy-Anteil ist insofern nicht zielführend, da das Ziel einer LNS ist, nur Teile der gefundenen Lösung zu verbessern, um somit schnell den Lösungsraum und die Nachbarschaften nach verbesserten Lösungen abzusuchen. Die Kapazität eines jeden Fahrzeuges kann ebenfalls eingestellt werden. Bei einer sehr kleinen Wahl kann es zu einer Fehlermeldung kommen, was bedeutet, dass die Kapazität zu knapp gewählt wurde und somit Nebenbedingungen des Modells verletzt wurden. Dies ist ein Signal, dass die Kapazität vergrößert werden sollte. Da die gesamte Nachfrage des Modells 532 beträgt, führt ein Kapazitätswert von 532 oder höher dazu, dass theoretisch ein Fahrzeug alle Kunden des Modells bedienen könnte. In diesem Fall liegt ein unbeschränktes Vehicle Routing Problem vor. Durch einen Schalter für die Wahl der Destroy-Operatoren können Random Ruin und Radial Ruin miteinander verglichen werden.
Die maximale Anzahl der Iterationen wurde in diesem Fall auf 100 festgelegt. Es bietet sich natürlich an, viel mehr als 100 Iterationen durchlaufen zu lassen oder alternativ mit einem Zeitimit zu arbeiten. Da jedoch im Rahmen dieser Arbeit vor allem die Veranschaulichung der LNS und dessen Grundzüge für eine Implementierung in einer Programmiersprache im Vordergrund stehen, ist es hier von Vorteil, die Ergebnisse nicht allzu groß und komplex werden zu lassen. Für eine didaktische Aufbereitung der LNS ist hierbei auch die grafische Visualisierung der Ergebnisse von zentraler Bedeutung und diese ist für eine große Anzahl an Iterationen durch lange Ladezeiten und zunehmender Komplexität nicht konzipiert. Außerdem haben die beobachteten Effekte natürlicherweise aufgrund der beschränkten Iterationszahl keine allgemeingültige Aussagekraft. Es lassen sich jedoch einige Tendenzen erkennen, welche Anstöße für weitere Nachforschungen im Rahmen einer größeren Rechenstudie geben können.
Bei der Wahl der Kapazität lässt sich beobachten, dass die Wahl einer kleineren Kapazität, unter der Voraussetzung, dass die Bildung des Tourenplans nicht primär durch die Kundennachfragen anstatt durch die verursachten Kosten der zurückgelegten Wegstrecke beeinflusst wird, zu bevorzugen ist. Falls die Kapazitätsrestriktion nicht zu stark einschränkend ist, kommt die LNS mit einer strengen bzw. starken Restriktion sehr gut zurecht. Dies belegt die Fähigkeit der LNS, sehr gut mit beschränkten Tourenplanungsproblemen umgehen zu können. Ein weiterer Aspekt stellt der Anteil der Zerstörung dar. Wie bereits gesagt, resultiert eine zu große Wahl dieses Destroy-Anteils in deutlich längeren Berechnungszeiten und grundsätlich auch in schlechteren Ergebnissen. Eine in jeder Iteration erneut erfolgende Durchführung eines komplexen Insertion-Prozesses ist folglich nicht effizient und somit auch nicht erstrebenswert. Ein Zerstörungsanteil von 0.2 wäre beispielweise eine gute Wahl, um bei vergleichsweise geringer Rechenzeit einen ausreichenden Teilbereich der aktuellen Lösung zu verändern. Außerdem lässt sich behaupten, dass die Verwendung des Radial Ruin-Operators oft bessere Ergebnisse liefert als die Verwendung eines Random Ruin-Operators. Dies liegt daran, dass ein Radial Ruin-Operator durch seinen Fokus auf einen zufälligen Kunden und dessen Nachbarschaft besser auf Problemzonen des Tourenplans einwirken kann, indem der komplette Bereich oder wesentliche Teile dieses Bereichs in einem Zug entfernt und im Anschluss neu zusammengefügt werden können. Ein besserer Fokus des Radial Ruin auf diese Bereiche führt zu einer schnelleren Verbesserung des Tourenplans und somit zu besseren Zielfunktionswerten. Ein Random Ruin-Operator wird womöglich, falls eine ausreichende Anzahl an Iterationen vorliegt und der Operator somit durch seine zufällige Auswahl der Knoten wahrscheinlich auch irgendwann zwangsläufig die Knoten der genannten Problemzonen entfernen wird, ebenfalls dieses Problem lösen. Jedoch ist davon auszugehen, dass dieser Operator dafür eben deutlich länger dafür brauchen wird, als ein Radial Ruin-Operator.
Bei der Wahl der Auswahlwahrscheinlichkeit einer verschlechternden Lösung ist zu beachten, dass hierbei keine allgemeingültige Aussage getroffen werden kann, ob eher eine hohe oder niedrige Wahrscheinlichkeit eingestellt werden sollte. Die Anpassung des Parameters dient vielmehr dazu, zu erkennen, ob das Verfahren in einem lokalen Optimum feststeckt. Nachdem eine Erhöhung des Parameters zu einer häufigeren Zulassung von verschlechternden Ergebnissen führt, kann anschließend der Effekt dieser Erhöhung auf den besten gefundenen Zielfunktionswert untersucht werden. Falls der ZF-Wert wirklich kleiner geworden ist, haben wir ein lokales Optimum entdeckt und gleichzeitig verlassen. Ob die neue Lösung nun ein globales Optimum darstellt, lässt sich allerdings nicht im Rahmen eines heuristischen Verfahrens feststellen. Es kann aber auch sein, dass eine Erhöhung der Auswahlwahrscheinlichkeit keinen verbessernden oder gar einen verschlechternden Effekt auf den besten gefundenen ZF-Wert hat. Entweder liegt hier keine Verbesserungsmöglichkeit vor, oder es sind noch weitere Iterationen nötig, um eine Verbesserung zu finden. Eine Wahl einer Wahrscheinlichkeit zwischen 0 und 1 könnte auch in manchen Situationen von Vorteil sein, da diese Kombination ermöglicht, sowohl die aktuelle Lösung tiefgehender untersuchen zu können, als auch den gesamten Lösungsraum breiter erforschen zu können. 

In [9]:
#CVRP
def DestroyAndRepairVRP(kapazität, f, iteration, auswahl, auswahlWKeit):
    #Startlösung
    tourenplan = copy.deepcopy(tourenplanSeed)
    kundenOffen = copy.deepcopy(kundenOffenSeed)
    lösungZulassen = None
    #Startlösung
    while kundenOffen != []:
        pOpt = M
        for j in kundenOffen:
            for k in fahrzeuge:
                if sum(nachfrage[i] for i in tourenplan[k]) + nachfrage[j] <= kapazität:
                    for m in range(1, len(tourenplan[k])):
                        if c[tourenplan[k][m-1],j] + c[j,tourenplan[k][m]] - c[tourenplan[k][m-1],tourenplan[k][m]] < pOpt:
                            kOpt = k
                            mOpt = m
                            jOpt = j
                            pOpt = c[tourenplan[k][m-1],j] + c[j,tourenplan[k][m]] - c[tourenplan[k][m-1], tourenplan[k][m]]
        tourenplan[kOpt].insert(mOpt,jOpt)
        kundenOffen.remove(jOpt)
    #Zielfunktion der Startlösung
    zf = 0
    for tour in tourenplan:
        for j in tour[:-1]:
            nachfolgerJ = tour[tour.index(j)+1]
            zf += c[j,nachfolgerJ]
    #Initialisierung der besten gefundenen Lösung mit ZF der Startlösung
    zfBest = zf
    tourenplanBest = tourenplan
    itBest = 0
    #Anzahl zufällig ausgewählter Knoten, die aus dem Tourenplan entfernt werden
    a = int(f * anzKunden)
    B = []
    #zur Nachvollziehbarkeit der Ergebnisse in anschließender Iterationsschleife
    random.seed(24)
    #Iterationen
    for it in range(iteration):
        #Auswahl des Destroy-Operators
        if auswahl == "Random Ruin":
            #Random-Ruin
            kundenInTour = copy.deepcopy(kunden)
            for i in range(a):
                entfernterKunde = random.choice(kundenInTour)
                kundenInTour.remove(entfernterKunde)
                B.append(entfernterKunde)
        else:
            #Radial-Ruin
            kundenInTour = copy.deepcopy(kunden)
            zufälligerKunde = random.choice(kundenInTour)
            nachbarschaftKunde = copy.deepcopy(c[zufälligerKunde,:])
            nachbarschaftKunde[0] = M
            for i in range(a):
                entfernterKunde = np.argmin(nachbarschaftKunde)
                nachbarschaftKunde[entfernterKunde] = M
                B.append(entfernterKunde)
        #Destroy
        for i in B:
            for tour in tourenplan:
                for j in tour:
                    if i == j:
                        tour.remove(i)
        #Repair
        while B != []:
            pOpt = M
            for j in B:
                for k in fahrzeuge:
                    if sum(nachfrage[i] for i in tourenplan[k]) + nachfrage[j] <= kapazität:
                        for m in range(1, len(tourenplan[k])):
                            if c[tourenplan[k][m-1],j] + c[j,tourenplan[k][m]] - c[tourenplan[k][m-1],tourenplan[k][m]] < pOpt:
                                kOpt = k
                                mOpt = m
                                jOpt = j
                                pOpt = c[tourenplan[k][m-1],j] + c[j,tourenplan[k][m]] - c[tourenplan[k][m-1], tourenplan[k][m]]
            tourenplan[kOpt].insert(mOpt,jOpt)
            B.remove(jOpt)
        #Ermittlung der Zielfunktion der aktuellen Iteration
        zf = 0
        for tour in tourenplan:
            for j in tour[:-1]:
                nachfolgerJ = tour[tour.index(j)+1]
                zf += c[j,nachfolgerJ]
        #Prüfen, ob keine Verbesserung vorliegt
        if zf >= zfBest:
            #Bestimmung mit festgelegter W'Keit, ob Verschlechterung akzeptiert wird
            anzahlJa = int(20 * auswahlWKeit)
            anzahlNein = int(20 * (1 - auswahlWKeit))
            listeWKeit = []
            for i in range(anzahlJa):
                listeWKeit.append("Ja")
            for i in range(anzahlNein):
                listeWKeit.append("Nein")
            lösungZulassen = random.choice(listeWKeit)
            if lösungZulassen == "Nein":
                tourenplan = copy.deepcopy(tourenplanBest)
                zf = copy.deepcopy(zfBest)
        #Prüfen, ob neuer bester ZF-Wert gefunden wurde
        if zf < zfBest:
            tourenplanBest = copy.deepcopy(tourenplan)
            zfBest = zf
            itBest = it + 1
            lösungZulassen = None
    #Plot
    for tour in tourenplan:
        x = []
        y = []
        for kunde in tour:
            x.append(koordinatenKnoten[kunde][0])
            y.append(koordinatenKnoten[kunde][1])
        plt.plot(x, y, 'r.-', 0, 0, 'bx')
        
    #Ausgabe des ZF-Wertes der aktuellen Iteration
    print("Lösung der aktuellen Iteration:")
    print("aktueller ZF-Wert:", zf)
    print("aktueller Tourenplan:", tourenplan)
    
    #Ausgabe des besten ZF-Wertes nach der aktuellen Iteration
    print("beste gefundene Lösung:")
    print("bester ZF-Wert:", zfBest)
    print("bester gefundener Tourenplan:", tourenplanBest)
    print("Iteration, in der die beste Lösung gefunden wurde:", itBest)
    if lösungZulassen == "Nein":
        print("Die aktuell gefundene Lösung wurde verworfen",
              "und es wird mit der bisher besten gefundenen Lösung weitergerechnet.")
    elif lösungZulassen == "Ja":
        print("Die aktuell gefundene Lösung wird trotz Verschlechterung beibehalten und zur weiteren Berechnung verwendet.")
    else:
        print("In der aktuellen Iteration wurde ein neuer bester Zielfunktionswert ermittelt.")

kapazitätSliderVRP = widgets.IntSlider(value=300, min=0, max=600, step=1, description="Kapazität:", continuous_update=False)
fractionSliderVRP = widgets.FloatSlider(value=0.2, min=0, max=1, step=0.05, description="Anteil der zu entfernenden Knoten:", 
                                        continuous_update=False)
iterationSliderVRP = widgets.IntSlider(value=100, min=0, max=100, step=1, description="Iteration:", 
                                       continuous_update=False)
auswahlDestroyVRP = widgets.ToggleButtons(options=["Random Ruin", "Radial Ruin"], description="Destroy-Operator:", 
                                          continuous_update=False)
verschlechterungZulassenVRP = widgets.FloatSlider(value=0, min=0, max=1, step=0.05, 
                                                  description="W'Keit, eine schlechtere Lösung zuzulassen:", 
                                                  continuous_update=False)

widgets.interact(DestroyAndRepairVRP, kapazität=kapazitätSliderVRP, f=fractionSliderVRP, iteration=iterationSliderVRP, 
                 auswahl=auswahlDestroyVRP, auswahlWKeit=verschlechterungZulassenVRP);

interactive(children=(IntSlider(value=300, continuous_update=False, description='Kapazität:', max=600), FloatS…

## Schluss
Im Rahmen dieser Arbeit wurden die Grundlagen der Large Neighbourhood Search-Metaheuristik erklärt und eine beispielhafte Implementierung in einer Programmiersprache vorgestellt. Hierbei stand die didaktische Aufbereitung des LNS-Verfahrens im Vordergrund, welche vor allem durch grafische Veranschaulichungen und der Möglichkeit, interaktiv auf die Ergebnisse und Berechnungen Einfluss zu nehmen, gestützt wurde. Abschließend lässt sich sagen, dass ein Jupyter Notebook für diese Art der Aufbereitung der LNS gut geeignet ist. Hierbei ist vor allen die Einbindung interaktiver Widgets, eine grafische Ausgabe der Ergebnisse und der Einbezug erklärender Textpassagen direkt überhalb der Code-Zeilen von Vorteil. Gleichwohl sollte beachtet werden, dass bei der Ausgabe der Code-Zeilen die Reihenfolge unbedingt beibehalten werden sollte. Hier ist eine Ausführung des Codes von oben nach unten erforderlich. Falls diese Reihenfolge nicht eingehalten wurde, sollte der Kernel neu gestartet und der Output der Code-Zeilen gelöscht werden, bevor das Notebook erneut von oben beginnend ausgeführt werden kann. Außerdem bietet die vorliegende Arbeit sehr viel Spielraum für Erweiterungen und zukünftige Untersuchungen. Es könnten beispielsweise noch weitere Nebenbedingungen wie kundenindividuelle Zeitfenster für die Bedienung der Nachfrage, weitere Destroy- oder auch Repair-Operatoren oder ein größeres Iterations- bzw. Berechnungszeitlimit hinzugefügt werden. Hierfür wäre auch insbesondere die Erstellung einer groß angelegten Rechenstudie, in welcher nicht die didaktische Aufbereitung, sondern eine analytische Untersuchung der Ergebnisse im Hauptfokus steht, interessant. 

## Quellen

+ Pisinger D. und S. Ropke (2010): Large Neighborhood Search. In: Gendreau M. und J.-Y. Potvin (Hrsg.): Handbook of Metaheuristics. International Series in Operations Research & Management Science, vol 146, Springer, Boston, S. 399-419
+ Schrimpf G.; J. Schneider, H. Stamm-Wilbrandt und G. Dueck (2000): Record Breaking Optimization Results Using the Ruin and Recreate Principle. In: Journal of Computational Physics, vol 159, S. 139-171
+ Shaw P. (1998): Using Constraint Programming and Local Search Methods to Solve Vehicle Routing Problems. In: Maher M. und JF. Puget (Hrsg.): Principles and Practice of Constraint Programming — CP98. Lecture Notes in Computer Science, vol 1520, Springer, Berlin, Heidelberg, S. 417-431
+ Skript zur Veranstaltung Logistics Management (Wintersemester 2019)